### PR TITLE
adding forward method for multihead attention

### PR DESCRIPTION
**Pull request by pmabbo13**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #1829
* #1828
* #1827
* #1826
* __->__ #1833
* #1832
* #1831
* #1830
* #1825

### CODE REVIEW

File: torchtext/prototype/t5/modules.py [hunk 1]
```diff
@@ -62,8 +64,340 @@ def __init__(
...
+            head_dim * self.num_heads == embed_dim
+        ), f"embed_dim {embed_dim} not divisible by num_heads {self.num_heads}"
+        # allow MHA to have different embedding dimensions when separate projection weights are used
+        assert (
+            key.shape[:2] == value.shape[:2]
+        ), f"key's sequence and batch dims {key.shape[:2]} do not match value's {value.shape[:2]}"
+
+        #
+        # compute in-projection
+        #
```
Comment (Nayef211) [inline torchtext/prototype/t5/modules.py:229]: nit: let's remove the extra lines in the comment

File: torchtext/prototype/t5/modules.py [hunk 2]
```diff
@@ -62,8 +64,340 @@ def __init__(
...
+            if key_padding_mask is not None:
+                key_padding_mask = F.pad(key_padding_mask, (0, 1))
+        else:
+            assert self.bias_k is None
+            assert self.bias_v is None
+
+        #
+        # reshape q, k, v for multihead attention and make em batch first
+        #
+
```
Comment (Nayef211) [inline torchtext/prototype/t5/modules.py:293]: nit: remove extra lines from comments

**Comment by Nayef211** [comment]
This is still a relatively large PR and it's a bit hard to tell which lines were added by you. It might be worthwhile leaving a comment on the PR highlighting the lines that are added here and are different from the original implementation (like you did in https://github.com/pytorch/text/pull/1812).

File: torchtext/prototype/t5/modules.py [hunk 3]
```diff
@@ -62,8 +64,340 @@ def __init__(
...
+            new_attn_mask = torch.zeros_like(attn_mask, dtype=q.dtype)
+            new_attn_mask.masked_fill_(attn_mask, float("-inf"))
+            attn_mask = new_attn_mask
+
+        # adjust dropout probability
+        if not self.training:
+            dropout_p = 0.0
+        else:
+            dropout_p = self.dropout
+
```
Comment (pmabbo13) [inline torchtext/prototype/t5/modules.py:356]: lines `357-377` encapsulate the changes made from _torch.nn.functional._multi_head_attention_forward_ to include relative attention bias. `position_bias` is then also added as a return value in lines `393, 400`.

**Comment by pmabbo13** [comment]
# Description

The forward method for T5MultiheadAttention is a modified version of [nn.Functional.multi_head_attention_forward](https://github.com/pytorch/pytorch/blob/5953fd9133c0bdcc0158acf1472fac403bc5f636/torch/nn/functional.py#L4909) meant to perform multihead attention with relative attention bias on the input query, key, and value tensors. The main modifications are as follows:

1. Parameters needed to compute relative attention bias are added as input arguments
2. Deprecated non-core functionalities such as `add_zero_attn` (adds a new batch of zeros to the key and value sequences at dim=1),  and adding bias terms to the key and value  projections
3. The nn.Functional.multi_head_attention_forward method reshapes the `q`, `k`, `v` to be 3D. Doing so appears to have led to discrepancies to the decoder outputs of the HF implementation when the input decoder sequence had a batch size larger than 1. The resolution was to shape these tensors to 4D (similarly to HF implementation), and that appeared to resolve the issue.

These changes are best visible via this [commit](https://github.com/pytorch/text/pull/1833/commits/f41597827e598ea934169003e0a03eefc5bc3168).

File: torchtext/prototype/t5/modules.py [hunk 4]
```diff
@@ -53,8 +54,244 @@ def __init__(
...
+                    )
+                attn_mask = attn_mask.view(1, 1, tgt_len, tgt_len).expand(bsz, self.num_heads, -1, -1)
+            else:
+                raise RuntimeError(f"attn_mask's dimension {attn_mask.dim()} is not supported")
+
+        # Prep key padding mask
+        if key_padding_mask is not None and key_padding_mask.dtype == torch.uint8:
+            warnings.warn("Byte tensor for key_padding_mask is not supported. Using bool tensor instead.")
+            key_padding_mask = key_padding_mask.to(torch.bool)
+
```
Comment (pmabbo13) [inline torchtext/prototype/t5/modules.py:224]: We were having issues from the original implementation where the q, k, v tensors were reshaped to `(batch_size * num_heads, seq_len, head_dim)`. This led to outputs that differed from the HF output when the input sequence to the decoder had batch size larger 1. The resolution to this was to shape these into 4D tensors of shape `(batch_size, num_heads, seq_len, head_dim)` which is the same reshaping done in the HF implementation.

File: torchtext/prototype/t5/modules.py [hunk 5]
```diff
@@ -53,8 +54,244 @@ def __init__(
...
+        key_padding_mask: Optional[Tensor] = None,
+        need_weights: bool = True,
+        attn_mask: Optional[Tensor] = None,
+        average_attn_weights: bool = False,
+        compute_relative_attention_bias=False,
+        relative_attention_num_buckets=32,
+        relative_attention_max_distance=128,
+        relative_attention_bias: Optional[Tensor] = None,
+        position_bias: Optional[Tensor] = None,
+    ) -> Tuple[Tensor, Optional[Tensor]]:
```
Comment (parmeet) [inline torchtext/prototype/t5/modules.py:71]: Looks like the return type doesn't match actual return. There are three variable being returned whereas annotation accounts for only 2?

File: torchtext/prototype/t5/modules.py [hunk 6]
```diff
@@ -53,8 +54,244 @@ def __init__(
...
+            position_bias=position_bias,
+            key_padding_mask=key_padding_mask,
+            need_weights=need_weights,
+            attn_mask=attn_mask,
+            average_attn_weights=average_attn_weights,
+        )
+        return attn_output, attn_output_weights, position_bias
+
+    # NOTE: Modified from https://github.com/pytorch/pytorch/blob/5953fd9133c0bdcc0158acf1472fac403bc5f636/torch/nn/functional.py#L4909
+    def t5_multi_head_attention_forward(
```
Comment (parmeet) [inline torchtext/prototype/t5/modules.py:140]: looks like this is a private function. can we precede with underscore?

File: torchtext/prototype/t5/modules.py [hunk 7]
```diff
@@ -53,8 +54,244 @@ def __init__(
...
+        compute_relative_attention_bias: bool,
+        relative_attention_num_buckets: Optional[int],
+        relative_attention_max_distance: Optional[int],
+        relative_attention_bias: Optional[Tensor],
+        position_bias: Optional[Tensor],
+        key_padding_mask: Optional[Tensor] = None,
+        need_weights: bool = True,
+        attn_mask: Optional[Tensor] = None,
+        average_attn_weights: bool = False,
+    ) -> Tuple[Tensor, Optional[Tensor]]:
```
Comment (parmeet) [inline torchtext/prototype/t5/modules.py:154]: same comment as above w.r.t return type.

File: torchtext/prototype/t5/modules.py [hunk 8]
```diff
@@ -53,8 +54,244 @@ def __init__(
...
+        attn_output, attn_output_weights = self._t5_dot_product_attention(q, k, v, position_bias, attn_mask, dropout_p)
+        attn_output = F.linear(attn_output, self.out_proj.weight, self.out_proj.bias)
+
+        if need_weights:
+            # Optionally average attention weights over heads
+            if average_attn_weights:
+                attn_output_weights = attn_output_weights.sum(dim=1) / self.num_heads
+
+            if not is_batched:
+                # Aqueeze the output if input was unbatched
```
Comment (parmeet) [inline torchtext/prototype/t5/modules.py:283]:
```suggestion
                # Squeeze the output if input was unbatched
```

File: torchtext/prototype/t5/modules.py [hunk 9]
```diff
@@ -53,8 +54,244 @@ def __init__(
...
+                attn_output = attn_output.squeeze(1)
+                attn_output_weights = attn_output_weights.squeeze(0)
+
+            return attn_output, attn_output_weights, position_bias
+
+        else:
+            if not is_batched:
+                # Aqueeze the output if input was unbatched
```
Comment (parmeet) [inline torchtext/prototype/t5/modules.py:291]:
```suggestion
                # Squeeze the output if input was unbatched
```

**Review by parmeet — APPROVED** [review]
Overall LGTM!

**Review by Nayef211 — APPROVED** [review]
LGTM, thanks for adding a detailed description with the changes wrt to original implementation and resolving all PR comments!
